### ISSUE TITLE
Density of a custom material raises zero despite knowing the density

### ISSUE THREAD

**Issue by drewj-usnctech** [terrapower/armi#1695]
Related to #1682: `Material.density` returns zero, probably under the assumption a subclass will override the method. And most cases that's true.

But, if you make a custom material in the blueprint file and provide a mass density, the `CustomMaterial` object does not override the density and will return zero. This is despite the fact that is known by the class

https://github.com/terrapower/armi/blob/7b26f79ac8c58b0477b0d7fa384b363d10fa9406/armi/materials/custom.py#L40

and used in the `pseudoDensity` method

https://github.com/terrapower/armi/blob/7b26f79ac8c58b0477b0d7fa384b363d10fa9406/armi/materials/custom.py#L42-L50

But, if you have a routine that does something with a subclass of `Material` and relies on `Material.density`, you can see strange problems because `CustomMaterial.density` falls back to `Material.density` which relies on the `refDens` property

https://github.com/terrapower/armi/blob/7b26f79ac8c58b0477b0d7fa384b363d10fa9406/armi/materials/material.py#L492-L514

which defaults to zero

https://github.com/terrapower/armi/blob/7b26f79ac8c58b0477b0d7fa384b363d10fa9406/armi/materials/material.py#L101

**Comment by john-science** [comment]
@drewj-usnctech Just to be sure. This features is meant to support "custom isotopics" in our blueprints:

https://github.com/terrapower/armi/blob/main/armi/reactor/blueprints/tests/test_componentBlueprint.py#L290-L293

Is that how you're using the `Custom(Material)` class?

**Comment by drewj-usnctech** [comment]
> @drewj-usnctech Just to be sure. This features is meant to support "custom isotopics" in our blueprints:
> 
> https://github.com/terrapower/armi/blob/7b26f79ac8c58b0477b0d7fa384b363d10fa9406/armi/reactor/blueprints/tests/test_componentBlueprint.py#L290-L293
> 
> Is that how you're using the `Custom(Material)` class?

That is correct

**Comment by john-science** [comment]
@drewj-tp I wanted to point you at an issue Mark opened up: https://github.com/terrapower/armi/issues/865

The idea for an ARMI material really is to set the reference density, and then have the math perturb it from there. So... is your request still to bypass the reference density for custom materials?

How would you like this to work, in an ideal world?

**Comment by john-science** [comment]
I fully agree that if you use the "custom isotopics" placeholders as materials they will totally break in a dozen ways. Pretty much any material method will return a stupid value: linear expansion, enthalpy, dynamicVisc, yieldStrength, etc etc.

It's certainly not just density.

There is no way around this, as the custom isotopics is not a full material definition. There just isn't enough information to fill in these gaps. If I add a mandatory reference density so you can get your density from this material. You still not have linear expansion, enthalpy, dynamicVisc, yieldStrength, etc etc.

**Comment by john-science** [comment]
I guess, since the linear expansion always returns zero, I can set `density` and `pseudoDensity` equal.

That is self-consistent.
